opty/utils.py

```diff
@@ -229,6 +230,140 @@ def parse_free(free, n, q, N):
+        The objective function to be minimized, which is a function of the
+        states and inputs.
+    state_symbols : iterable of symbols
+        The state variables.
+    input_symbols : iterable of symbols
+        The input variables.
+    unknown_symbols : iterable of symbols
+        The unknown parameters.
+    N : int
+        Number of collocation nodes, i.e. the number of time steps.
+    node_time_interval : float
+        The value of the time interval. The default is 1.0, as this term only
+        appears in the objective function as a scaling factor.
+    integration_method : str, optional
+        The method used to integrate the system. The default is "backward
+        euler".
+
+    """
+    def lambdify_function(expr, multiplication_array, take_sum):
+        if take_sum:
+            def integration_function(x):
+                return node_time_interval * np.sum(x * multiplication_array)
+        else:
+            def integration_function(x):
+                return node_time_interval * x * multiplication_array
+        return sm.lambdify(
+            (states, inputs, params), expr,
+            modules=[{int_placeholder.name: integration_function}, "numpy"],
+            cse=True)
+
+    # Parse function arguments
+    states = sm.ImmutableMatrix(state_symbols)
+    inputs = sm.ImmutableMatrix(sort_sympy(input_symbols))
+    params = sm.ImmutableMatrix(sort_sympy(unknown_symbols))
+    if states.shape[1] > 1 or inputs.shape[1] > 1 or params.shape[1] > 1:
+        raise ValueError(
+            'The state, input, and unknown symbols must be column matrices.')
+    n, q = states.shape[0], inputs.shape[0]
+    i_idx, r_idx = n * N, (n + q) * N
+
+    # Compute analytical gradient of the objective function
+    objective_grad = sm.ImmutableMatrix([objective]).jacobian(
+        states[:] + inputs[:] + params[:])
+
+    # Replace the integral with a custom function
+    int_placeholder = sm.Function("_IntegralFunction")
+    wild_expr = sm.Wild("expr")
+    repl = (sm.Integral(wild_expr, (me.dynamicsymbols._t)),
+            int_placeholder(wild_expr))
+    objective = objective.replace(*repl)
+    objective_grad = tuple(objective_grad.replace(*repl))
+
+    # Replace zeros with an array of zeros, otherwise lambdify will return a
+    # scalar zero instead of an array of zeros.
+    objective_grad = tuple(
+        np.zeros(N) if grad == 0 else grad for grad in objective_grad[:n + q]
+        ) + tuple(objective_grad[n + q:])
+
+    # Define evaluation functions based on the integration method
+    if integration_method == "backward euler":
+        obj_expr_eval = lambdify_function(
+            objective, np.hstack((0, np.ones(N - 1))), True)
+        obj_grad_time_expr_eval = lambdify_function(
+            objective_grad[:n + q], np.hstack((0, np.ones(N - 1))), False)
+        obj_grad_param_expr_eval = lambdify_function(
+            objective_grad[n + q:], np.hstack((0, np.ones(N - 1))), True)
```

Comment:
I don't follow these lines perfectly. What is the reason to force the sum or not?

Reply:
Yeah, it all feels a bit like a cheat.
So, the multiplications with arrays make sure you always have the right shape and they also give weights to the nodes. E.g. the first node doesn't do anything in backward Euler.
For the gradient, we in general of course do not want to sum the states, as we actually want to compute the separate pieces. However, if you for some reason got an integral over a state or input as the gradient of an unknown, then we do need to just integrate it, i.e. take the sum.

P.S. The explanation is a bit messy. I can try to make it more clear.